README.md
#One Page Scroll by Pete R.
Create an Apple-like one page scroll website (iPhone 5S website) with One Page Scroll plugin
Created by [Pete R.](http://www.thepetedesign.com), Founder of [BucketListly](http://www.bucketlistly.com)

License: [Attribution-ShareAlike 4.0 International](http://creativecommons.org/licenses/by-sa/4.0/deed.en_US)

This is forked from Pete R's  [One Page Scroll](https://github.com/peachananr/onepage-scroll), and provides more functionality and bug fixes.

See: [Change Log](https://github.com/leftstick/onepage-scroll/blob/master/ChangeLog.md)


[![One Page Scroll](http://www.thepetedesign.com/images/onepage_scroll_image.png "One Page Scroll")](http://www.thepetedesign.com/demos/onepage_scroll_demo.html)

## Requirement

jQuery (1.9.0 or later)

note: jQuery 1.9.0 or later is strongly recommended because using jQuery less than 1.8.3(or greater than 2.1.0) and jquery.onepage-scroll.js together turns out to be a hash-based XSS vulnerabiliry.

Try it: http://plnkr.co/edit/U3Tscez35JathRQYlAjz?p=preview

## Install

```shell
bower install git://github.com/leftstick/onepage-scroll.git#1.4.0 -S
```

## Demo
[View demo](http://www.thepetedesign.com/demos/onepage_scroll_demo.html)

## Compatibility
Modern browsers such as Chrome, Firefox, and Safari on both desktop and smartphones have been tested. Not tested on IE.

## Basic Usage
One Page Scroll let you transform your website into a one page scroll website that allows users to scroll one page at a time. It is perfect for creating a website in which you want to present something to the viewers. For example, [Apple's iPhone 5S website](http://www.apple.com/iphone-5s/) uses the same technique.


To add this to your website, simply include the latest jQuery library together with `jquery.onepage-scroll.js`, `onepage-scroll.css` into your document's `<head>` and call the function as follows:

````html
<body>
  ...
  <div class="main">
    <section>...</section>
    <section>...</section>
    ...
  </div>
  ...
</body>
````
Container "Main" must be one level below the `body` tag in order to make it work full page. Now call the function to activate as follows:
 
````javascript
$(".main").onepage_scroll({
   sectionContainer: "section",     // sectionContainer accepts any kind of selector in case you don't want to use section
   easing: "ease",                  // Easing options accepts the CSS3 easing animation such "ease", "linear", "ease-in", 
                                    // "ease-out", "ease-in-out", or even cubic bezier value such as "cubic-bezier(0.175, 0.885, 0.420, 1.310)"
   animationTime: 1000,             // AnimationTime let you define how long each section takes to animate
   pagination: true,                // You can either show or hide the pagination. Toggle true for show, false for hide.
   updateURL: false,                // Toggle this true if you want the URL to be updated automatically when the user scroll to each page.
   beforeMove: function(index) {},  // This option accepts a callback function. The function will be called before the page moves.
   afterMove: function(index) {},   // This option accepts a callback function. The function will be called after the page moves.
   loop: false,                     // You can have the page loop back to the top/bottom when the user navigates at up/down on the first/last page.
   keyboard: true,                  // You can activate the keyboard controls
   responsiveFallback: false,        // You can fallback to normal page scroll by defining the width of the browser in which
                                    // you want the responsive fallback to be triggered. For example, set this to 600 and whenever 
                                    // the browser's width is less than 600, the fallback will kick in.
   responsiveHeightFallback: false, // you can fallback to normal page scroll by defining the height of the browser.
   disableMouseMove: false,          // You can have the mouse wheel functionality disabled by setting this attribute to true. The default value is false.
   moveUpKeys: [33, 38]             // This option allow you to specify key code for move up operation. NavigateUp/PageUp by default.
   moveDownKeys: [34, 40]             // This option allow you to specify key code for move down operation. NavigateDown/PageDown by default.
   backtoTopKeys: [36]             // This option allow you to specify key code for move to top operation. Home by default.
   gotoBottom: [35]             // This option allow you to specify key code for move to bottom operation. End by default.

});
````
And that's it. Now, your website should work the same way Apple's iPhone 5S website does. You should be able to swipe up/down as well (thanks to [Eike Send](https://github.com/eikes) for his swipe events!) when viewing your website on mobile phones.

## Public Methods
You can also trigger page move programmatically as well:

### $.fn.destroy_onepage_scroll()
This method allows you to unload the `onepage-scroll` functionality. This might be useful when you are developing SPA.

````javascript
  $(".main").destroy_onepage_scroll(options);
````
>Note: Only `sectionContainer` is required in options.

### $.fn.moveUp()
This method allows you to move the page up by one. This action is equivalent to scrolling up/swiping down.

````javascript
  $(".main").moveUp();
````

### $.fn.moveDown()
This method allows you to move the page down by one. This action is equivalent to scrolling down/swiping up.

````javascript
  $(".main").moveDown();
````

### $.fn.moveTo(page_index)
This method allows you to move to the specified page index programatically.

````javascript
  $(".main").moveTo(3);
````

## Callbacks
You can use callbacks to perform actions before or after the page move.

### beforeMove(next_page_index, next_section_element)
This callback gets called before the plugin performs its move.

````javascript
  $(".main").onepage_scroll({
    beforeMove: function(index, next_el) {
      ...
    }
  });
````

### afterMove(next_page_index, next_section_element)
This callback gets called after the move animation was performed.

````javascript
  $(".main").onepage_scroll({
    afterMove: function(index, next_el) {
      ...
    }
  });
````

If you want to see more of my plugins, visit [The Pete Design](http://www.thepetedesign.com/#design), or follow me on [Twitter](http://www.twitter.com/peachananr) and [Github](http://www.github.com/peachananr).

## Other Resources
- [OnePageScroll.js: Creating an Apple’s iPhone 5S Website](http://www.onextrapixel.com/2013/09/18/onepagescroll-js-creating-an-apples-iphone-5s-website/)
- [Eike Send's jQuery Swipe Events](https://github.com/eikes/jquery.swipe-events.js)
- [CSS Easing generator by Matthew Lein](http://matthewlein.com/ceaser/)


ChangeLog.md
## v1.4.0 ##

- [**New Feature**]: Add vertical response fallback, Since the view looks weird while someone working with his/her monitor as portrait. A blank area shown at bottom of the screen.

## v1.3.6 ~ 1.3.9 ##

- [**Bug fix**]: swipe gesture doesn't work properly on touchable devices such as ipad.

## v1.3.5 ##

- [**Bug fix**]: replace the disable_onepage_scroll to destroy_onepage_scroll, since it's not accurate

## v1.3.4 ##

- [**New Feature**]: move up/down, top/bottom operations could be specified by the user

## v1.3.3 ##

- [**New Feature**]: add disable mouse wheel option and make the scrollbar enabled within one section while the section is bigger than the screen

## v1.3.2 ##

- [**Bug fix**]: window.resize listener was not unbinded while disable the onepage-scroll


## v1.3.0 ##

- [**Bug fix**]: active functionality worked incorrect while the location.hash is come up with other values. 
- [**Improvement**]: add gulp as build system tool.



Gruntfile.js
/*global module:false*/
module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    // Metadata.
    pkg: grunt.file.readJSON('bower.json'),
    // Task configuration.
    concat: {
      options: {
        stripBanners: false
      },
      dist: {
        src: ['lib/<%= pkg.name %>.js'],
        dest: 'dist/<%= pkg.name %>.js'
      }
    },
    cssmin: {
      dist: {
        files: {
          'dist/<%= pkg.name %>.css': ['lib/**/*.css']
        }
      }
    },
    uglify: {
      dist: {
        src: '<%= concat.dist.dest %>',
        dest: 'dist/<%= pkg.name %>.min.js'
      }
    },
    copy: {
      demo: {
        expand: true,
        cwd: 'dist/',
        src: ['**'],
        dest: 'Demo/'
      }
    },
    jshint: {
      options: {
        curly: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        newcap: true,
        noarg: true,
        sub: true,
        undef: true,
        unused: true,
        boss: true,
        eqnull: true,
        browser: true,
        globals: {
          jQuery: true
        }
      },
      lib_test: {
        src: ['app/scripts/*.js']
      }
    },
    watch: {
      gruntfile: {
        files: '<%= jshint.gruntfile.src %>',
        tasks: ['jshint:gruntfile']
      },
      lib_test: {
        files: '<%= jshint.lib_test.src %>',
        tasks: ['jshint:lib_test']
      }
    }
  });

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-concat');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-cssmin');
  grunt.loadNpmTasks('grunt-contrib-copy');

  // Default task.
  grunt.registerTask('default', ['jshint', 'concat', 'uglify', 'cssmin', 'copy']);

};


jquery.onepage-scroll.js
/* ===========================================================
 * jquery-onepage-scroll.js v1.4.0
 * ===========================================================
 * Copyright 2013 Pete Rojwongsuriya.
 * http://www.thepetedesign.com
 *
 * Create an Apple-like website that let user scroll
 * one page at a time
 *
 * Credit: Eike Send for the awesome swipe event
 * https://github.com/leftstick/onepage-scroll
 *
 * License: GPL v3
 *
 * ========================================================== */

(function($) {

    var defaults = {
        sectionContainer: "section",
        easing: "ease",
        animationTime: 1000,
        pagination: true,
        updateURL: false,
        keyboard: true,
        beforeMove: null,
        afterMove: null,
        loop: false,
        responsiveFallback: false,
        responsiveHeightFallback: false,
        disableMouseMove: false,
        moveUpKeys: [33, 38],
        moveDownKeys: [34, 40],
        backtoTopKeys: [36],
        gotoBottom: [35]
    };

    /*------------------------------------------------*/
    /*  Credit: Eike Send for the awesome swipe event */
    /*------------------------------------------------*/

    $.fn.swipeEvents = function() {
        return this.each(function() {

            var startX,
                startY,
                $this = $(this);

            $this.bind('touchstart', touchstart);

            function touchstart(event) {
                var touches = event.originalEvent.touches;
                if (touches && touches.length) {
                    startX = touches[0].pageX;
                    startY = touches[0].pageY;
                    $this.bind('touchmove', touchmove);
                }
            }

            function touchmove(event) {
                var touches = event.originalEvent.touches;
                if (touches && touches.length) {
                    var deltaX = startX - touches[0].pageX;
                    var deltaY = startY - touches[0].pageY;

                    if (deltaX >= 50) {
                        $this.trigger("swipeLeft");
                    }
                    if (deltaX <= -50) {
                        $this.trigger("swipeRight");
                    }
                    if (deltaY >= 50) {
                        $this.trigger("swipeUp");
                    }
                    if (deltaY <= -50) {
                        $this.trigger("swipeDown");
                    }
                    if (Math.abs(deltaX) >= 50 || Math.abs(deltaY) >= 50) {
                        $this.unbind('touchmove', touchmove);
                    }
                    event.preventDefault();
                }
            }

        });
    };


    $.fn.onepage_scroll = function(options) {
        var settings = $.extend({}, defaults, options),
            el = $(this),
            sections = $(settings.sectionContainer),
            total = sections.length,
            status = "off",
            topPos = 0,
            lastAnimation = 0,
            quietPeriod = 500,
            paginationList = "";

        $.fn.transformPage = function(settings, pos, index, next_el) {
            if (typeof settings.beforeMove == 'function') settings.beforeMove(index, next_el);
            $(this).css({
                "-webkit-transform": "translate3d(0, " + pos + "%, 0)",
                "-webkit-transition": "-webkit-transform " + settings.animationTime + "ms " + settings.easing,
                "-moz-transform": "translate3d(0, " + pos + "%, 0)",
                "-moz-transition": "-moz-transform " + settings.animationTime + "ms " + settings.easing,
                "-ms-transform": "translate3d(0, " + pos + "%, 0)",
                "-ms-transition": "-ms-transform " + settings.animationTime + "ms " + settings.easing,
                "transform": "translate3d(0, " + pos + "%, 0)",
                "transition": "transform " + settings.animationTime + "ms " + settings.easing
            });

            $(this).one('webkitTransitionEnd otransitionend oTransitionEnd msTransitionEnd transitionend', function(e) {
                if (typeof settings.afterMove == 'function') settings.afterMove(index, next_el);
            });
        };

        $.fn.moveDown = function() {
            var el = $(this),
                index = $(settings.sectionContainer + ".active").data("index");
            current = $(settings.sectionContainer + "[data-index='" + index + "']");
            next = $(settings.sectionContainer + "[data-index='" + (index + 1) + "']");
            if (next.length < 1) {
                if (settings.loop === true) {
                    pos = 0;
                    next = $(settings.sectionContainer + "[data-index='1']");
                } else {
                    return;
                }

            } else {
                pos = (index * 100) * -1;
            }
            current.removeClass("active");
            next.addClass("active");
            if (settings.pagination === true) {
                $(".onepage-pagination li a" + "[data-index='" + index + "']").removeClass("active");
                $(".onepage-pagination li a" + "[data-index='" + next.data("index") + "']").addClass("active");
            }

            $("body")[0].className = $("body")[0].className.replace(/\bviewing-page-\d.*?\b/g, '');
            $("body").addClass("viewing-page-" + next.data("index"));

            if (history.replaceState && settings.updateURL === true) {
                var href = window.location.href.substr(0, window.location.href.indexOf('#')) + "#" + (index + 1);
                history.pushState({}, document.title, href);
            }
            el.transformPage(settings, pos, next.data("index"), next);
        };

        $.fn.moveUp = function() {
            var el = $(this),
                index = $(settings.sectionContainer + ".active").data("index");
            current = $(settings.sectionContainer + "[data-index='" + index + "']");
            next = $(settings.sectionContainer + "[data-index='" + (index - 1) + "']");

            if (next.length < 1) {
                if (settings.loop === true) {
                    pos = ((total - 1) * 100) * -1;
                    next = $(settings.sectionContainer + "[data-index='" + total + "']");
                } else {
                    return;
                }
            } else {
                pos = ((next.data("index") - 1) * 100) * -1;
            }
            current.removeClass("active");
            next.addClass("active");
            if (settings.pagination === true) {
                $(".onepage-pagination li a" + "[data-index='" + index + "']").removeClass("active");
                $(".onepage-pagination li a" + "[data-index='" + next.data("index") + "']").addClass("active");
            }
            $("body")[0].className = $("body")[0].className.replace(/\bviewing-page-\d.*?\b/g, '');
            $("body").addClass("viewing-page-" + next.data("index"));

            if (history.replaceState && settings.updateURL === true) {
                var href = window.location.href.substr(0, window.location.href.indexOf('#')) + "#" + (index - 1);
                history.pushState({}, document.title, href);
            }
            el.transformPage(settings, pos, next.data("index"), next);
        };

        $.fn.moveTo = function(page_index) {
            current = $(settings.sectionContainer + ".active");
            next = $(settings.sectionContainer + "[data-index='" + (page_index) + "']");
            if (next.length > 0) {
                current.removeClass("active");
                next.addClass("active");
                $(".onepage-pagination li a" + ".active").removeClass("active");
                $(".onepage-pagination li a" + "[data-index='" + (page_index) + "']").addClass("active");
                $("body")[0].className = $("body")[0].className.replace(/\bviewing-page-\d.*?\b/g, '');
                $("body").addClass("viewing-page-" + next.data("index"));

                pos = ((page_index - 1) * 100) * -1;

                if (history.replaceState && settings.updateURL === true) {
                    var href = window.location.href.substr(0, window.location.href.indexOf('#')) + "#" + (page_index - 1);
                    history.pushState({}, document.title, href);
                }
                el.transformPage(settings, pos, page_index, next);
            }
        };

        function responsive() {
            if ((settings.responsiveFallback && $(window).width() < settings.responsiveFallback) || (settings.responsiveHeightFallback && $(window).height() > settings.responsiveHeightFallback)) {
                $("body").addClass("disabled-onepage-scroll");
                $(document).unbind('mousewheel DOMMouseScroll');
                el.swipeEvents().unbind("swipeDown swipeUp touchstart touchmove");
            } else {
                if ($("body").hasClass("disabled-onepage-scroll")) {
                    $("body").removeClass("disabled-onepage-scroll");
                    $("html, body, .wrapper").animate({
                        scrollTop: 0
                    }, "fast");
                }


                el.swipeEvents().bind("swipeDown", function(event) {
                        if (!$("body").hasClass("disabled-onepage-scroll")) event.preventDefault();
                        el.moveUp();
                    }).bind("swipeUp", function(event) {
                    if (!$("body").hasClass("disabled-onepage-scroll")) event.preventDefault();
                    el.moveDown();
                });

                if (!settings.disableMouseMove) {

                    $(document).bind('mousewheel DOMMouseScroll', function(event) {
                        // event.preventDefault();
                        current = $(settings.sectionContainer + ".active");
                        var delta = event.originalEvent.wheelDelta || -event.originalEvent.detail;
                        if (current.height() >= current.find('.page-container').height() || (current.scrollTop() === 0 && delta >= 0) || ((current[0].scrollHeight - current.scrollTop() === current.height()) && delta < 0)) {
                            event.preventDefault();
                            init_scroll(event, delta);
                        }
                    });
                }
            }
        }


        function init_scroll(event, delta) {
            var deltaOfInterest = delta,
                timeNow = new Date().getTime();
            // Cancel scroll if currently animating or within quiet period
            if (timeNow - lastAnimation < quietPeriod + settings.animationTime) {
                event.preventDefault();
                return;
            }

            if (deltaOfInterest < 0) {
                el.moveDown();
            } else {
                el.moveUp();
            }
            lastAnimation = timeNow;
        }

        // Prepare everything before binding wheel scroll

        $('html').css('overflow', 'hidden').css('height', '100%');
        $('body').css('overflow', 'hidden').css('height', '100%');

        el.addClass("onepage-wrapper").css("position", "relative");
        $.each(sections, function(i) {
            $(this).addClass("ops-section").attr("data-index", i + 1);
            topPos = topPos + 100;
            if (settings.pagination === true) {
                paginationList += "<li><a data-index='" + (i + 1) + "' href='#" + (i + 1) + "'></a></li>";
            }
        });

        // Create Pagination and Display Them
        if (settings.pagination === true) {
            $("<ul class='onepage-pagination'>" + paginationList + "</ul>").prependTo("body");
            posTop = (el.find(".onepage-pagination").height() / 2) * -1;
            el.find(".onepage-pagination").css("margin-top", posTop);
        }

        if (window.location.hash !== "" && window.location.hash !== "#1" && $(settings.sectionContainer + "[data-index='" + window.location.hash.replace("#", "") + "']").length > 0) {
            init_index = window.location.hash.replace("#", "");
            $(settings.sectionContainer + "[data-index='" + init_index + "']").addClass("active");
            $("body").addClass("viewing-page-" + init_index);
            if (settings.pagination === true) $(".onepage-pagination li a" + "[data-index='" + init_index + "']").addClass("active");

            next = $(settings.sectionContainer + "[data-index='" + (init_index) + "']");
            if (next) {
                next.addClass("active");
                if (settings.pagination === true) $(".onepage-pagination li a" + "[data-index='" + (init_index) + "']").addClass("active");
                $("body")[0].className = $("body")[0].className.replace(/\bviewing-page-\d.*?\b/g, '');
                $("body").addClass("viewing-page-" + next.data("index"));
                if (history.replaceState && settings.updateURL === true) {
                    var href = window.location.href.substr(0, window.location.href.indexOf('#')) + "#" + (init_index);
                    history.pushState({}, document.title, href);
                }
            }
            pos = ((init_index - 1) * 100) * -1;
            el.transformPage(settings, pos, init_index);

        } else {
            $(settings.sectionContainer + "[data-index='1']").addClass("active");
            $("body").addClass("viewing-page-1");
            if (settings.pagination === true) $(".onepage-pagination li a" + "[data-index='1']").addClass("active");
        }
        if (settings.pagination === true) {
            $(".onepage-pagination li a").click(function() {
                var page_index = $(this).data("index");
                el.moveTo(page_index);
            });
        }

        if (!settings.disableMouseMove) {
            $(document).bind('mousewheel DOMMouseScroll', function(event) {

                current = $(settings.sectionContainer + ".active");
                var delta = event.originalEvent.wheelDelta || -event.originalEvent.detail;
                if (current.height() >= current.find('.page-container').height() || (current.scrollTop() === 0 && delta >= 0) || ((current[0].scrollHeight - current.scrollTop() === current.height()) && delta < 0)) {
                    event.preventDefault();
                    if (!$("body").hasClass("disabled-onepage-scroll")) {
                        init_scroll(event, delta);
                    }
                }
            });
        }



        if (settings.responsiveFallback || settings.responsiveHeightFallback) {
            $(window).resize(function() {
                responsive();
            });

            responsive();
        }

        if (settings.keyboard === true) {
            $(document).keydown(function(e) {
                var tag = e.target.tagName.toLowerCase();

                if (!$("body").hasClass("disabled-onepage-scroll") && (tag !== 'input' && tag !== 'textarea')) {

                    if ($.inArray(e.which, settings.moveUpKeys) > -1) {
                        el.moveUp();
                    } else if ($.inArray(e.which, settings.moveDownKeys) > -1) {
                        el.moveDown();
                    } else if ($.inArray(e.which, settings.backtoTopKeys) > -1) {
                        el.moveTo(1);
                    } else if ($.inArray(e.which, settings.gotoBottom) > -1) {
                        var lastIndex = sections.last().data("index");
                        el.moveTo(lastIndex);
                    }
                }

            });
        }
        return false;
    };

    $.fn.destroy_onepage_scroll = function(options) {
        var settings = $.extend({}, defaults, options);
        var el = $(this);
        var sections = $(settings.sectionContainer);

        $('html').css('overflow', '').css('height', '');
        $('body').css('overflow', '').css('height', '');

        el.removeClass("onepage-wrapper");
        $.each(sections, function(i) {
            $(this).removeClass("ops-section active").removeAttr("data-index");
        });

        el.swipeEvents().unbind("swipeDown swipeUp touchstart touchmove");
        $("body").removeClass("disabled-onepage-scroll");
        $('.onepage-pagination li a').unbind('click');
        $('ul.onepage-pagination').remove();

        var classListOnBody = $('body').attr('class').split(/\s+/);
        $.each(classListOnBody, function(index, item) {
            if (item.indexOf('viewing-page-') >= 0) {
                $('body').removeClass(item);
            }
        });

        $(document).unbind('mousewheel DOMMouseScroll');
        $(window).unbind('resize');
        $(document).unbind('keydown');
    };


})(window.jQuery);